onadata/libs/filters.py

```diff
@@ -189,7 +189,21 @@ def filter_queryset(self, request, queryset, view):
         """Filter by ``XForm.shared_data = True`` for anonymous users."""
         if request.user.is_anonymous:
             return queryset.filter(Q(shared_data=True))
-        return queryset
+
+        # Get all xforms from the queryset
+        xform_ids = list(set(queryset.values_list("id", flat=True)))
+        allowed_xform_ids = []
+
+        # Check permissions for each xform
+        for xform_id in xform_ids:
```

Comment:
Why not loop through the `queryset`  instead of loop through `xform_ids`? This will allow us to also utilize `queryset_iterator`